File: script/apps/Aries/Creator/Game/GUI/EditDialog.js
angular.module('DialogItem', [])
.controller('DialogItemController', function ($scope, $http) {
    $scope.filename = (location.search.split('filename=')[1] || '').split('&')[0];

    $scope.actions = {
        accept: "",
        gotonext: "",
    };

    // json data source
    $scope.data = {
        title: "",
        gossips: [],
        triggers: [],
        quests:[],
    };

    $scope.itemTypes = ["item", "virtualitem"];
    $scope.buttonActions = ["gotonext", "accept", "goto item_name", "close"];
    $scope.avatarNames = ["player"];

    // add a new gossip
    $scope.addGossip = function () {
        var dialog = [];
        $scope.addDialogitem(dialog);
        $scope.data.gossips.push(dialog);
    };
    $scope.removeItem = function (fromItems, index, minCount) {
        if (fromItems.length > minCount)
            fromItems.splice(index, 1);
        else
            alert("at least " + minCount + " item must exist");
    };

    // add gossip item in a gossip
    $scope.addDialogitem = function (dialog) {
        dialog.push({
            name: "",
            avatar: { name: "", text: "" },
            content: "",
            buttons: [],
        });
    };

    $scope.addDialogButton = function (item) {
        if (item.buttons == null)
            item.buttons = [];
        item.buttons.push({});
    }

    // shuffel items up
    $scope.moveItemUp = function (items, index) {
        if (index > 0) {
            var tmp = items[index - 1];
            items[index - 1] = items[index];
            items[index] = tmp;
        }
    }

    $scope.addTrigger = function () {
        var trigger = {
            input: [],
            dialog: [],
            output: [],
        };
        $scope.addDialogitem(trigger.dialog);
        $scope.data.triggers.push(trigger);
    };

    $scope.addRuleitem = function (rule) {
        rule.push({
            name: "virtualitem",
            id: "",
            count: 1,
        });
    };

    $scope.reload = function () {
        var url = "ajax/open?file=script/apps/Aries/Creator/Game/GUI/EditDialog.page&filename="
            + encodeURIComponent($scope.filename) + "&action=get_dialog_file";
        $http.get(url).then(function (response) {
            $scope.data = response.data;
            $scope.last_message = "file loaded";
        });
    }

    $scope.save = function () {
        // var url = "ajax/open" + window.location.href.slice(window.location.href.indexOf('?')) + "&action=save_dialog_file";
        var url = "ajax/open?file=script/apps/Aries/Creator/Game/GUI/EditDialog.page&filename="
                    + encodeURIComponent($scope.filename) + "&action=save_dialog_file";

        $http.post(url, $scope.data).then(function (response) {
            if (response.data.success) {
                $scope.last_message = "file is successfully saved to " + response.data.filename;
            }
            else {
                $scope.last_message = "failed to saved to " + response.data.filename;
            }
        });
    }

    if ($scope.filename != null) {
        $scope.reload();
    }
    if (Page)
        Page.ShowSideBar(false);
});

File: script/apps/Aries/Creator/Game/GUI/EditDialog.lua
--[[
Title: Edit Item Dialog
Author(s): LiXizhi
Date: 2016/4/5
Desc: 
use the lib:
------------------------------------------------------------
NPL.load("(gl)script/apps/Aries/Creator/Game/GUI/EditDialog.lua");
local EditDialog = commonlib.gettable("MyCompany.Aries.Game.GUI.EditDialog");
EditDialog.ShowPage(itemStack, filename)
-------------------------------------------------------
]]
local ItemClient = commonlib.gettable("MyCompany.Aries.Game.Items.ItemClient");
local EntityManager = commonlib.gettable("MyCompany.Aries.Game.EntityManager");

local EditDialog = commonlib.gettable("MyCompany.Aries.Game.GUI.EditDialog");

function EditDialog.OnInit()
end

function EditDialog.ShowPage(itemStack, filename, triggerEntity)
	filename = filename or itemStack:GetDataField("tooltip");
	local url = "/open npl://open?file=script/apps/Aries/Creator/Game/GUI/EditDialog.page&title=Edit Dialog";
	if(filename and filename~="") then
		url = url.."&filename="..filename;
	end
	GameLogic.RunCommand(url);
end
